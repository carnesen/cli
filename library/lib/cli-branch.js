"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CliBranch = void 0;
const constants_1 = require("./constants");
/**
 * A factory function for creating command branches
 * @remarks
 * @param options
 * @returns Returns the newly-created branch object
 * @example
 * ```typescript
 * const cloudBranch = CliBranch()
 * ```
 */
function CliBranch(options) {
    return {
        ...options,
        commandType: constants_1.CLI_BRANCH,
    };
}
exports.CliBranch = CliBranch;
//# sourceMappingURL=cli-branch.js.map