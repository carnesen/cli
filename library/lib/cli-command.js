"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CliCommand = void 0;
const constants_1 = require("./constants");
/**
 *
 * A factory function for creating CLI command commands
 * @param options
 * @returns The newly-created `command`.
 */
function CliCommand(options) {
    return {
        ...options,
        commandType: constants_1.CLI_COMMAND,
    };
}
exports.CliCommand = CliCommand;
//# sourceMappingURL=cli-command.js.map