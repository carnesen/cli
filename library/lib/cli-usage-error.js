"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CliUsageError = exports.CLI_USAGE_ERROR = void 0;
exports.CLI_USAGE_ERROR = 'CLI_USAGE_ERROR';
/**
 * A custom error with code {@linkcode CLI_USAGE_ERROR}
 * @remarks
 * {@linkcode runCliAndExit} catches this class of error and prints command-line usage
 * (i.e. "Usage: cloud users lists ...") for the current command. If the `message`
 * property is provided, `runCliAndExit` will normal `Error`. See also print `Error: <your
 * message>`. See also {@linkcode CliTerseError}.
 * @example
 * ```typescript
 * throw new CliUsageError("Expected one or more values")
 * ```
 */
class CliUsageError extends Error {
    /**
     *
     * @param message An `Error` message
     * @param commandStack Used internally for argument processing
     */
    constructor(message, commandStack) {
        super(message);
        Object.defineProperty(this, "code", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "commandStack", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.code = exports.CLI_USAGE_ERROR;
        this.commandStack = commandStack;
        Object.setPrototypeOf(this, new.target.prototype);
    }
}
exports.CliUsageError = CliUsageError;
//# sourceMappingURL=cli-usage-error.js.map