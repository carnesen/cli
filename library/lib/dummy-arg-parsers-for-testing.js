"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.dummyRequiredArgParser = exports.dummyArgParser = exports.DUMMY_ARG_PARSER_THROW_NON_TRUTHY = exports.DUMMY_ARG_PARSER_THROW = exports.DUMMY_ARG_PARSER_THROWN_INTENTIONALLY = exports.DUMMY_ARG_PARSER_EMPTY_ARRAY_WAS_PASSED = exports.DUMMY_ARG_PARSER_UNDEFINED_WAS_PASSED = void 0;
exports.DUMMY_ARG_PARSER_UNDEFINED_WAS_PASSED = 'undefined was passed';
exports.DUMMY_ARG_PARSER_EMPTY_ARRAY_WAS_PASSED = 'undefined was passed';
exports.DUMMY_ARG_PARSER_THROWN_INTENTIONALLY = 'thrown intentionally';
exports.DUMMY_ARG_PARSER_THROW = 'throw';
exports.DUMMY_ARG_PARSER_THROW_NON_TRUTHY = 'throw-non-truthy';
exports.dummyArgParser = {
    placeholder: '',
    parse(args) {
        if (typeof args === 'undefined') {
            return exports.DUMMY_ARG_PARSER_UNDEFINED_WAS_PASSED;
        }
        if (args.length === 0) {
            return exports.DUMMY_ARG_PARSER_EMPTY_ARRAY_WAS_PASSED;
        }
        if (args[0] === exports.DUMMY_ARG_PARSER_THROW) {
            throw new Error(exports.DUMMY_ARG_PARSER_THROWN_INTENTIONALLY);
        }
        if (args[0] === exports.DUMMY_ARG_PARSER_THROW_NON_TRUTHY) {
            // eslint-disable-next-line no-throw-literal
            throw '';
        }
        return args[0];
    },
};
exports.dummyRequiredArgParser = {
    placeholder: '<foo>',
    required: true,
    parse: exports.dummyArgParser.parse,
};
//# sourceMappingURL=dummy-arg-parsers-for-testing.js.map