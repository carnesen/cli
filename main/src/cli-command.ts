import { CLI_COMMAND } from './constants';
import {
  AnyArgParser,
  AnyNamedArgParsers,
  ValueFromArgParser,
  NamedValues,
  ICliArgParser,
} from './cli-arg-parser';

export type CliCommandOptions<
  TPositionalArgParser extends AnyArgParser,
  TNamedArgParsers extends AnyNamedArgParsers,
  TEscapedArgParser extends AnyArgParser
> = {
  /**
   * Name used in command-line usage
   */
  name: string;
  /**
   * The function that defines your command logic.
   *
   * @param positionalValue
   * @param namedValues
   * @param escapedValue
   * @returns A value (synchronously) or a `Promise` that resolves to a value
   * @remarks
   * By default, {@linkcode runCliAndExit} `console.log`s the value returned by `action`.
   * The type of the `args` argument received by `action` is derived by the `args`
   * property of the command. Similarly, the `options` argument type is derived from
   * `command.options`.
   *
   */
  action: (
    positionalValue: ValueFromArgParser<TPositionalArgParser>,
    namedValues: NamedValues<TNamedArgParsers>,
    escapedValue: ValueFromArgParser<TEscapedArgParser>,
  ) => any;
  positionalArgParser?: TPositionalArgParser;
  namedArgParsers?: TNamedArgParsers;
  escapedArgParser?: TEscapedArgParser;
  /**
   * A string that will be included in `Usage:` if present.
   */
  description?: string;
  /**
   * If true, don't show this command in usage docs.
   */
  hidden?: boolean;
};

/**
 * Interface describing a "command" command
 * @typeParam TPositionalArgParser Type of the "positional" arguments parser
 * @typeParam TNamedArgParsers Type of the ""
 */
export interface ICliCommand<
  TPositionalArgParser extends AnyArgParser,
  TNamedArgParsers extends AnyNamedArgParsers,
  TEscapedArgParser extends AnyArgParser
> extends CliCommandOptions<TPositionalArgParser, TNamedArgParsers, TEscapedArgParser> {
  commandType: typeof CLI_COMMAND;
}

/**
 *
 * A factory function for creating CLI command commands
 * @param options
 * @returns The newly-created `command`.
 */
export function CliCommand<
  TPositional extends AnyArgParser = ICliArgParser<undefined, false>,
  TNamed extends AnyNamedArgParsers = any,
  TEscaped extends AnyArgParser = ICliArgParser<undefined, false>
>(
  options: CliCommandOptions<TPositional, TNamed, TEscaped>,
): ICliCommand<TPositional, TNamed, TEscaped> {
  return {
    ...options,
    commandType: CLI_COMMAND,
  };
}
