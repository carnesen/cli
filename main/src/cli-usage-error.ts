import { Node } from './cli-node';

/** "code" of a {@link CliUsageError} */
export const CLI_USAGE_ERROR = 'CLI_USAGE_ERROR';

/**
 * Thrown to print command usage and an error message but not a `stack`
 */
export class CliUsageError extends Error {
  /** The string constant {@linkcode CLI_USAGE_ERROR} */
  public readonly code: typeof CLI_USAGE_ERROR;

  /** Used internally for constructing the command-line usage string */
  public node?: Node;

  /**
   * @param message If provided, {@linkcode runCliAndExit} will also print "Error: \<your
   * message\>"
   * @param node Used internally for constructing the command-line usage string
   */
  constructor(message?: string, node?: Node) {
    super(message);
    this.code = CLI_USAGE_ERROR;
    this.node = node;
    Object.setPrototypeOf(this, new.target.prototype);
  }
}
