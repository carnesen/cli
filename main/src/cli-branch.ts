import { ICliCommand } from './cli-command';

/** "kind" of an {@linkcode ICliBranch} */
export const CLI_BRANCH = 'CLI_BRANCH';

/** Options for {@linkcode CliBranch} */
export interface ICliBranchOptions {
  /** Name of this command branch, typically a category like "user" or "cloud" */
  name: string;

  /** A short description of this branch for command-line usage */
  description?: string;

  /** If `true`, this branch will not appear in command-line usage */
  hidden?: boolean;

  /** Branches and/or commands underneath this branch */
  children: (ICliBranch | ICliCommand<any, any, any>)[];
}

/**
 * A branch in a command tree
 */
export interface ICliBranch extends ICliBranchOptions {
  /** Used internally for discriminating between {@linkcode ICliBranch}'s and
   * {@linkcode ICliCommand}'s */
  kind: typeof CLI_BRANCH;
}

/**
 * A factory for {@linkcode ICliBranch} objects
 * @param options
 * @returns An {@linkcode ICliBranch} object
 * @example
 * ```typescript
 * const cloudBranch = CliBranch()
 * ```
 */
export function CliBranch(options: ICliBranchOptions): ICliBranch {
  return {
    ...options,
    kind: CLI_BRANCH,
  };
}
