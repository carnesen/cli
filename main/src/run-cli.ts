import { partitionArgs } from './partition-args';
import { getNamedValues } from './get-named-values';
import { CliUsageError, CLI_USAGE_ERROR } from './cli-usage-error';
import { CliTerseError } from './cli-terse-error';
import { findVersion } from './find-version';
import { parseArgs } from './parse-args';
import { navigateToCommand } from './navigate-to-command';
import { ICliBranch } from './cli-branch';
import { ICliCommand } from './cli-command';

/** A JavaScript command runner. Useful for unit testing CLI logic. */
export interface IRunCli {
  (...args: string[]): Promise<any>;
}

/** (Advanced) Wrap the default {@linkcode IRunCli} to modify its behavior */
export interface ICliEnhancer {
  (runCli: IRunCli): IRunCli;
}

export type RunCliOptions = { enhancer?: ICliEnhancer };

/**
 * A factory for {@linkcode IRunCli}s, the core of {@linkcode runCliAndExit}
 *
 * @param root The root of this CLI's command tree
 */
export function RunCli(
  root: ICliBranch | ICliCommand<any, any, any>,
  options: RunCliOptions = {},
): IRunCli {
  const { enhancer } = options;

  if (enhancer) {
    return enhancer(runCli);
  }

  return runCli;

  async function runCli(...args: string[]) {
    // If the very first argument is --version, return this software's version identifier.
    if (args[0] === '--version') {
      const version = await findVersion();
      if (!version) {
        throw new CliTerseError('Failed to find a "version" string');
      }
      return version;
    }

    const [locationInCommandTree, remainingArgs] = navigateToCommand(root, args);

    const { positionalArgs, namedArgs, escapedArgs } = partitionArgs(remainingArgs);
    if (namedArgs.help) {
      throw new CliUsageError(undefined, locationInCommandTree);
    }
    const command = locationInCommandTree.current;
    let argsValue: any;
    if (command.positionalParser) {
      // Note that for named and escaped args, we distinguish between
      // `undefined` and `[]`. For example, "cli" gives an escaped args
      // `undefined` whereas "cli --" gives an escaped args `[]`. For the
      // "positionalArgs", however, there is no such distinction. By convention,
      // we elect here to pass in `undefined` rather than an empty array when no
      // positional arguments are passed.
      argsValue = await parseArgs(
        command.positionalParser,
        positionalArgs.length > 0 ? positionalArgs : undefined,
        undefined,
        locationInCommandTree,
      );
    } else if (positionalArgs.length > 0) {
      throw new CliUsageError(
        `Unexpected argument "${positionalArgs[0]}" : Command "${command.name}" does not accept positional arguments`,
        locationInCommandTree,
      );
    }

    const namedValues = await getNamedValues(
      command.namedParsers || {},
      namedArgs,
      locationInCommandTree,
    );

    let escapedValue: any;
    if (command.escapedParser) {
      escapedValue = await parseArgs(
        command.escapedParser,
        escapedArgs,
        '--',
        locationInCommandTree,
      );
    } else if (escapedArgs) {
      throw new CliUsageError(
        `Command "${command.name}" does not allow "--" as an argument`,
      );
    }

    try {
      const result = await command.action(argsValue, namedValues, escapedValue);
      return result;
    } catch (exception) {
      if (exception && exception.code === CLI_USAGE_ERROR) {
        exception.locationInCommandTree = locationInCommandTree;
      }
      throw exception;
    }
  }
}
