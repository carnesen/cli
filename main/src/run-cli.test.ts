import { runAndCatch } from '@carnesen/run-and-catch';
import { CliBranch } from './cli-branch';
import { CliCommand } from './cli-command';
import { dummyArgParser } from './dummy-arg-parsers-for-testing';
import { RunCli, CliEnhancer } from './run-cli';
import { findVersion } from './find-version';
import { CLI_USAGE_ERROR } from './cli-usage-error';

const commandWithNamedArgParsers = CliCommand({
  name: 'command-with-named-args',
  namedArgParsers: {
    foo: dummyArgParser,
  },
  action(...args) {
    return args;
  },
});

const commandWithPositionalArgParser = CliCommand({
  name: 'command-with-positional-args',
  positionalArgParser: dummyArgParser,
  action(...args) {
    return args;
  },
});

const commandWithEscapedArgParser = CliCommand({
  name: 'command-with-escaped-arg-parser',
  escapedArgParser: dummyArgParser,
  action(...args) {
    return args;
  },
});

const root = CliBranch({
  name: 'cli',
  children: [
    commandWithPositionalArgParser,
    commandWithNamedArgParsers,
    commandWithEscapedArgParser,
  ],
});

const cliArgRunner = RunCli(root);

describe(RunCli.name, () => {
  it('calls the enhancer if provided', async () => {
    const spy = jest.fn();
    const enhancer: CliEnhancer = (innerArgRunner) => async (...args: string[]) => {
      spy(...args);
      await innerArgRunner(...args);
    };
    const enhancedArgRunner = RunCli(commandWithPositionalArgParser, {
      enhancer,
    });
    await enhancedArgRunner('foo', 'bar');
    expect(spy.mock.calls).toEqual([['foo', 'bar']]);
  });

  it('returns version string from package.json if "--version" is passed', async () => {
    const version = await findVersion();
    expect(await cliArgRunner('--version')).toBe(version);
  });

  it('throws USAGE error with empty message if --help is passed', async () => {
    const exception = await runAndCatch(cliArgRunner, '--help');
    expect(exception.code).toBe(CLI_USAGE_ERROR);
    expect(exception.message).toBeFalsy();
  });

  it('throws USAGE error with empty message if last command is a branch and no additional args is present', async () => {
    const exception = await runAndCatch(cliArgRunner);
    expect(exception.code).toBe(CLI_USAGE_ERROR);
    expect(exception.message).toBeFalsy();
  });

  it('throws USAGE error "bad command" if last command is a branch and additional args is present', async () => {
    const exception = await runAndCatch(cliArgRunner, 'oops');
    expect(exception.code).toBe(CLI_USAGE_ERROR);
    expect(exception.message).toMatch(/bad command/i);
    expect(exception.message).toMatch('"oops"');
    expect(exception.message).toMatchSnapshot();
  });

  it('throws USAGE error "positional arguments" if last command is a command without positionalArgParser property and additional args is present', async () => {
    const exception = await runAndCatch(
      cliArgRunner,
      commandWithNamedArgParsers.name,
      'oops',
    );
    expect(exception.code).toBe(CLI_USAGE_ERROR);
    expect(exception.message).toMatch('Unexpected argument "oops"');
    expect(exception.message).toMatch(commandWithNamedArgParsers.name);
    expect(exception.message).toMatch(/positional arguments/i);
    expect(exception.message).toMatchSnapshot();
  });

  it('Passes parsed positional value as first argument of the "action" function', async () => {
    const positionalArgs = ['foo', 'bar'];
    const result = await cliArgRunner(
      commandWithPositionalArgParser.name,
      ...positionalArgs,
    );
    expect(result).toEqual([dummyArgParser.parse(positionalArgs), {}, undefined]);
  });

  it('Passes parsed named values as second argument of the "action" function', async () => {
    const namedArgs = ['--foo', 'bar'];
    const result = await cliArgRunner(commandWithNamedArgParsers.name, ...namedArgs);
    expect(result).toEqual([
      undefined,
      { foo: dummyArgParser.parse(['bar']) },
      undefined,
    ]);
  });

  it(`Throws USAGE error 'does not allow "--"' if command does not have an "escaped" property`, async () => {
    const exception = await runAndCatch(
      cliArgRunner,
      commandWithPositionalArgParser.name,
      '--',
    );
    expect(exception.code).toBe(CLI_USAGE_ERROR);
    expect(exception.message).toMatch(commandWithPositionalArgParser.name);
    expect(exception.message).toMatch('does not allow "--"');
  });

  it('Passes parsed escaped value as third argument of the "action" function', async () => {
    const result = await cliArgRunner(commandWithEscapedArgParser.name, '--');
    expect(result).toEqual([
      undefined,
      {},
      commandWithEscapedArgParser.escapedArgParser!.parse([]),
    ]);
  });
});
