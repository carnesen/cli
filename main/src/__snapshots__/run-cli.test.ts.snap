// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RunCli throws USAGE error "bad command" if last command is a branch and additional args is present 1`] = `"Bad command \\"oops\\""`;

exports[`RunCli throws USAGE error "positional arguments" if last command is a command without positionalValuedParser property and additional args is present 1`] = `"Unexpected argument \\"oops\\" : Command \\"command-with-named-args\\" does not accept positional arguments"`;
